cache/image_cache.go

```diff
@@ -0,0 +1,90 @@
+package cache
+
+import (
+	"encoding/json"
+	"io"
+	"log"
+
+	"github.com/buildpack/lifecycle"
+	"github.com/buildpack/lifecycle/image"
+	"github.com/pkg/errors"
+)
+
+//go:generate mockgen -package testmock -destination testmock/image_factory.go github.com/buildpack/lifecycle/cache ImageFactory
+
+type ImageFactory interface {
+	NewEmptyLocal(string) image.Image
+}
+
+type ImageCache struct {
+	logger    *log.Logger
+	factory   ImageFactory
+	origImage image.Image
+	newImage  *image.LoggingImage
+}
+
+func NewImageCache(logger *log.Logger, factory ImageFactory, origImage image.Image) *ImageCache {
+	newImage := factory.NewEmptyLocal(origImage.Name())
+	return &ImageCache{
+		logger:    logger,
+		factory:   factory,
+		origImage: origImage,
+		newImage:  image.NewLoggingImage(logger, newImage),
+	}
+}
+
+func (c *ImageCache) Name() string {
+	return c.origImage.Name()
+}
+
+func (c *ImageCache) SetMetadata(metadata lifecycle.CacheMetadata) error {
+	data, err := json.Marshal(metadata)
+	if err != nil {
+		return errors.Wrap(err, "marshall metadata")
+	}
+	return c.newImage.SetLabel(lifecycle.CacheMetadataLabel, string(data))
+}
+
+func (c *ImageCache) RetrieveMetadata() (lifecycle.CacheMetadata, bool, error) {
```

Comment:
Is this `bool` really useful? Is `false` + empty struct ever any different from `true` + empty struct? If the distinction is necessary, name the parameters so it's clear what the `bool` is for.

Reply:
@sclevine Yeah we considered this. Ultimately we landed here because there didn't seem to be an idiomatic way to check for this value on the caller side, and using pointers also felt weird. We felt a bool was clearest. Will opt to name it if you're fine with that (unless you have a better idea).

Reply:
@sclevine Ended up removing the bool. You're right that it wasn't super necessary

Reply:
Done.